File: src/1049-Last-Stone-Weight-II/1049.js
var lastStoneWeightII = function(stones) {
    let sum = 0;
    for (let s of stones) sum += s;
    let mem = new Array((sum>>1) + 1);
    mem[0] = 1;
    
    for (let i of stones) {
        for (let j = sum>>1; j >= i; --j) {
            mem[j] |= mem[j - i];
        }
    }
    for (let i = sum>>1; i > 0; --i) {
        if (mem[i]) return sum - 2*i;
    }
    return 0;
};

File: src/1049-Last-Stone-Weight-II/1049.py
class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        all_sum = sum(stones)
        mem = [0]*(all_sum//2+1)
        mem[0] = 1
        for i in stones:
            for j in range(all_sum//2, i-1, -1):
                mem[j] |= mem[j - i]

        for i in range(all_sum//2, -1, -1):
            if mem[i]:
                return all_sum - 2*i
        return 0